app/models/agents/attribute_difference_agent.rb

```diff
@@ -0,0 +1,104 @@
+      the difference or change of a specific attribute in comparison to the previous
+      event received.
+
+      `attribute` specifies the name of the attribute to be used from the event.
+
+      `method` specifies if it should be...
+
+      * `percentage_change` eg. Previous value was 160, new value is 116. Percentage change is `-27.5`
+      * `decimal_difference` eg. Previous value was 5.5, new value is 15.2. Difference is `9.7`
+      * `integer_difference` eg. Previous value was 50, new value is 40. Difference is `-10`
+
+      `expected_update_period_in_days` is used to determine if the Agent is working.
+
+      The resulting event will be a copy of the received event with the difference
+      or change added as an extra attribute. If you use the `percentage_change` the
+      attribute will be formatted as such `{{attribute}}_change`, otherwise it will
+      be `{{attribute}}_diff`.
+    MD
+
+    event_description <<-MD
+      This will change based on the source event.
+    MD
+
+    def default_options
+      {
+        'attribute' => 'rate',
+        'method' => 'integer_difference',
+        'expected_update_period_in_days' => 1
+      }
+    end
+
+    def validate_options
+      unless options['attribute'].present? && options['method'].present? && options['expected_update_period_in_days'].present?
+        errors.add(:base, 'The attribute, method and expected_update_period_in_days fields are all required.')
+      end
+    end
+
+    def working?
+      event_created_within?(interpolated['expected_update_period_in_days']) && !recent_error_logs?
+    end
+
+    def receive(incoming_events)
+      incoming_events.each do |event|
+        handle(interpolated(event), event)
+      end
+    end
+
+    private
+
+    def handle(opts, event = nil)
+      attribute = opts['attribute']
+      attribute_value = event.payload[attribute].nil? ? 0 : event.payload[attribute]
+      payload = event.payload.dup
+
+      if opts['method'] == 'percentage_change'
+        change = calculate_percentage_change(attribute_value)
+        payload["#{attribute}_change"] = change
+
+      elsif opts['method'] == 'decimal_difference'
+        difference = calculate_decimal_difference(attribute_value)
+        payload["#{attribute}_diff"] = difference
+
+      elsif opts['method'] == 'integer_difference'
+        difference = calculate_integer_difference(attribute_value)
+        payload["#{attribute}_diff"] = difference
+      end
+
+      created_event = create_event(payload: payload)
+      log('Propagating new event', outbound_event: created_event, inbound_event: event)
+      update_memory(attribute_value)
+    end
+
+    def calculate_integer_difference(new_value)
+      return 0 if last_value.nil?
+      (new_value.to_i - last_value.to_i)
+    end
+
+    def calculate_decimal_difference(new_value)
+      return 0.0 if last_value.nil?
+      (BigDecimal.new(new_value) - BigDecimal.new(last_value)).to_f
```

Comment:
It is the same, but you may want to force an integer difference. Eg. your input data was decimal, but you don't care and just want integers back out, or if you only want to deal in whole numbers. Thought i'd just add it as an extra option.